server/datastore/mysql/hosts.go

```diff
@@ -1648,11 +1652,63 @@ WHERE
 }
 
 func (ds *Datastore) UpdateOSVersions(ctx context.Context) error {
+	counts, err := ds.gatherOSCountsDB(ctx)
+	if err != nil {
+		return ctxerr.Wrapf(ctx, err, "gathering counts stats for os versions")
+	}
+
+	teams, err := ds.TeamsSummary(ctx)
+	if err != nil {
+		return ctxerr.Wrapf(ctx, err, "getting team summary")
+	}
+
+	countsByTeamId := make(map[uint]*json.RawMessage)
+	for _, c := range *counts {
+		countsByTeamId[c.TeamID] = c.OSVersions
+	}
+
+	for _, t := range teams {
+		if countsByTeamId[t.ID] == nil {
+			var raw json.RawMessage
+			raw, err = json.Marshal([]fleet.OSVersion{})
+			if err != nil {
+				return err
+			}
+			countsByTeamId[t.ID] = &raw
+		}
+	}
+
+	var insertArgs []interface{}
+	for id, counts := range countsByTeamId {
+		insertArgs = append(insertArgs, id, "os_versions", counts)
+	}
+
+	insertValues := strings.TrimSuffix(strings.Repeat("(?, ?, ?),", len(countsByTeamId)), ",")
+
+	sql := fmt.Sprintf(`
+INSERT INTO aggregated_stats (id, type, json_value) 
+VALUES %s 
+ON DUPLICATE KEY UPDATE
+	json_value = VALUES(json_value),
+	updated_at = CURRENT_TIMESTAMP
+`, insertValues)
+
+	if _, err := ds.writer.ExecContext(ctx, sql, insertArgs...); err != nil {
+		return ctxerr.Wrap(ctx, err, "update aggregated stats for os versions")
+	}
+
+	return nil
+}
+
+type osCounts struct {
+	TeamID     uint             `db:"id"`
+	OSVersions *json.RawMessage `db:"json_value"`
+}
+
+func (ds *Datastore) gatherOSCountsDB(ctx context.Context) (*[]osCounts, error) {
```

Comment:
I believe there's no need to return a pointer to a slice, you can just return `[]osCounts` (a slice can be `nil`).